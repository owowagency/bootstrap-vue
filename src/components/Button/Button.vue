<template>
    <button
        class="btn"
        :class="classes"
        :disabled="disabled"
    >
        <slot />
    </button>
</template>

<script lang="ts">
import useDisabled, {disabledProps} from '@/composables/useDisabled';
import useSize, {sizeProps} from '@/composables/useSize';
import useVariant, {variantProps} from '@/composables/useVariant';
import {computed} from 'vue';
import useClasses from '@/composables/useClasses';

export const buttonProps = {
    ...disabledProps,
    outline: {
        type: Boolean,
        default: false,
    },
    ...sizeProps,
    ...variantProps,
};
</script>

<script lang="ts" setup>
const props = defineProps(buttonProps);

const variantClassTemplate = computed(() => `btn-${props.outline ? 'outline-' : ''}{0}`);

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    useSize(props.size, 'btn-{0}').sizeClass.value,
    useVariant(props.variant, variantClassTemplate).variantClass.value,
]));
</script>
