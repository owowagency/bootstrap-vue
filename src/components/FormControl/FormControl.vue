<template>
    <input
        ref="input"
        :autofocus="autofocus"
        class="form-control"
        :class="classes"
        :disabled="disabled"
        :placeholder="placeholder"
        :readonly="readonly"
        :type="type"
        :value="modelValue"
        @input="emit('update:modelValue', $event.target.value)"
    >
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import {formControlProps} from '@/composables/useFormControl';
import useClasses from '@/composables/useClasses';
import useDisabled from '@/composables/useDisabled';
import useSize from '@/composables/useSize';
</script>

<script lang="ts" setup>
const props = defineProps(formControlProps);

const emit = defineEmits(['update:modelValue']);

const input = ref<HTMLInputElement>();

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    props.plainText ? 'form-control-plaintext' : '',
    props.readonly ? 'readonly' : '',
    useSize(props.size, 'form-control-{0}').sizeClass.value,
]));

defineExpose({input});
</script>
