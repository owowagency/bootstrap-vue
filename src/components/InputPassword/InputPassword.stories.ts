import InputPassword from '.';
import {ref} from 'vue';
import {sizes} from '@/composables/useSize';

export default {
    title: 'Forms/Input password',
    component: InputPassword,
    argTypes: {
        size: {
            control: {type: 'select'},
            options: sizes,
        },
    },
};

const template = (args) => ({
    components: {InputPassword},
    setup: () => ({args, value: ref('test')}),
    template: '<InputPassword v-model="value" v-bind="args" />',
});

export const Default = template.bind({});

export const Sizing = (args) => ({
    components: {InputPassword},
    setup() {
        return {args, sizes};
    },
    template: '<InputPassword v-for="s of sizes" class="mb-1" :size="s" v-bind="args">',
});

export const Rules = template.bind({});

Rules.args = {
    rules: [
        {
            rule: /\d/,
            message: 'One number',
        },
        {
            rule: /[a-z]/,
            message: 'One lowercase character',
        },
        {
            rule: /[A-Z]/,
            message: 'One upper character',
        },
        {
            rule: /.{7,}/,
            message: '7 characters',
        },
    ],
};
