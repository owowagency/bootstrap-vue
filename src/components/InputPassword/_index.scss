.input-password-icon {
    vertical-align: $spinner-vertical-align;
}

.input-password-feedback {
    list-style: none;
    margin-top: $form-feedback-margin-top;
    margin-bottom: 0;
    padding-left: 0;
    
    @each $state, $data in $form-validation-states {
        .is-#{$state} {
            position: relative;
            padding-left: calc(1em + .5rem);
            background-repeat: no-repeat;
            background-position: 0;
            background-size: 1em;

            color: map-get($data, 'color');
            background-image: escape-svg(map-get($data, 'icon'));
        }
    }
}
