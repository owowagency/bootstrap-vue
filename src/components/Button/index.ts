import Button from './Button.vue';

export * from './Button.vue';
export default Button;
