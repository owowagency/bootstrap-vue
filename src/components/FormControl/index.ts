import FormControl from './FormControl.vue';

export default FormControl;
