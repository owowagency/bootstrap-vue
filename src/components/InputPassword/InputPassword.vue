<template>
    <InputGroup :size="props.size">
        <template #append>
            <Button
                v-bind="propsButton"
                @click="toggleType"
            >
                <svg
                    v-if="type === 'password'"
                    xmlns="http://www.w3.org/2000/svg"
                    class="input-password-icon"
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 16 16"
                >
                    <path d="M16 8s-3-5.5-8-5.5S0 8 0 8s3 5.5 8 5.5S16 8 16 8zM1.173 8a13.133 13.133 0 0 1 1.66-2.043C4.12 4.668 5.88 3.5 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.133 13.133 0 0 1 14.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755C11.879 11.332 10.119 12.5 8 12.5c-2.12 0-3.879-1.168-5.168-2.457A13.134 13.134 0 0 1 1.172 8z" />
                    <path d="M8 5.5a2.5 2.5 0 1 0 0 5 2.5 2.5 0 0 0 0-5zM4.5 8a3.5 3.5 0 1 1 7 0 3.5 3.5 0 0 1-7 0z" />
                </svg>

                <svg
                    v-else
                    xmlns="http://www.w3.org/2000/svg"
                    class="input-password-icon"
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 16 16"
                >
                    <path d="M13.359 11.238C15.06 9.72 16 8 16 8s-3-5.5-8-5.5a7.028 7.028 0 0 0-2.79.588l.77.771A5.944 5.944 0 0 1 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.134 13.134 0 0 1 14.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755-.165.165-.337.328-.517.486l.708.709z" />
                    <path d="M11.297 9.176a3.5 3.5 0 0 0-4.474-4.474l.823.823a2.5 2.5 0 0 1 2.829 2.829l.822.822zm-2.943 1.299.822.822a3.5 3.5 0 0 1-4.474-4.474l.823.823a2.5 2.5 0 0 0 2.829 2.829z" />
                    <path d="M3.35 5.47c-.18.16-.353.322-.518.487A13.134 13.134 0 0 0 1.172 8l.195.288c.335.48.83 1.12 1.465 1.755C4.121 11.332 5.881 12.5 8 12.5c.716 0 1.39-.133 2.02-.36l.77.772A7.029 7.029 0 0 1 8 13.5C3 13.5 0 8 0 8s.939-1.721 2.641-3.238l.708.709zm10.296 8.884-12-12 .708-.708 12 12-.708.708z" />
                </svg>
            </Button>
        </template>

        <FormControl
            v-bind="propsFormControl"
            :type="type"
            @update:model-value="emit('update:modelValue', $event)"
        />
    </InputGroup>
</template>

<script lang="ts">
import Button, {buttonProps} from '@/components/Button';
import FormControl from '@/components/FormControl';
import InputGroup from '@/components/InputGroup';
import extractKeysFrom from '@/library/extractKeysFrom';
import {formControlProps} from '@/composables/useFormControl';
import {ref} from 'vue';
import {sizeProps} from '@/composables/useSize';
</script>

<script lang="ts" setup>
const props = defineProps({
    ...buttonProps,
    ...formControlProps,
    ...sizeProps,
});

const emit = defineEmits<{
    (e: 'update:modelValue', value?: string): void;
}>();

const propsFormControl = extractKeysFrom(Object.keys(formControlProps), props);

const propsButton = extractKeysFrom(Object.keys(buttonProps), props);

const type = ref<'password'|'text'>('password');

const toggleType = () => type.value = type.value === 'password' ? 'text' : 'password';
</script>
