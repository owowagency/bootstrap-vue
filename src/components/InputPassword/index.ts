import InputPassword from './InputPassword.vue';

export default InputPassword;
