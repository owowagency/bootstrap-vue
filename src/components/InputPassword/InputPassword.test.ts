import {mount, shallowMount} from '@vue/test-utils';
import {nextTick, ref} from 'vue';
import Button from '@/components/Button';
import FormControl from '@/components/FormControl';
import InputPassword from '.';

describe('template', () => {
    componentRenderTest(InputPassword, {}, false);

    componentRenderTest(
        InputPassword,
        {
            props: {
                modelValue: 'Password',
                rules: [
                    {rule: /Password/, feedback: 'Yoooo'},
                    {rule: /Drowssap/, feedback: 'Ooooy'},
                ],
            },
        },
        false,
        'renders rules feedback',
    );
});

describe('toggleType', () => {
    it('toggles type', async() => {
        const wrapper = mount(InputPassword);

        const button = wrapper.findComponent(Button);

        const formControl = wrapper.findComponent(FormControl);

        expect(formControl.props('type')).toBe('password');

        await button.trigger('click');

        expect(formControl.props('type')).toBe('text');

        await button.trigger('click');

        expect(formControl.props('type')).toBe('password');
    });
});

describe('isValid', () => {
    const rules = [
        {rule: /[a-zA-Z]/, feedback: 'alpha'},
        {rule: /\d/, feedback: 'number'},
    ];

    it('is valid when all rules are valid', () => {
        const wrapper = shallowMount(
            InputPassword,
            {
                props: {
                    modelValue: 'A1',
                    rules,
                },
            },
        );

        expect(wrapper.vm.isValid).toBe(true);
    });

    it('is not valid when not all rules are valid', () => {
        const wrapper = shallowMount(
            InputPassword,
            {
                props: {
                    modelValue: 'A',
                    rules,
                },
            },
        );

        expect(wrapper.vm.isValid).toBe(false);
    });

    it('emits on change', async() => {
        const modelValue = ref('A');

        const wrapper = mount(InputPassword, {props: {modelValue, rules}});

        modelValue.value = 'A1';

        await nextTick();

        expect(wrapper.emitted('update:isValid')).toEqual([[true]]);
    });
});
