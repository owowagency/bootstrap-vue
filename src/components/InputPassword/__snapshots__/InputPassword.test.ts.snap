// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `"<input-group-stub size=\\"md\\"></input-group-stub>"`;
